### PR TITLE
Update Hold/Tap Actions to Objects

### PR DESCRIPTION
**Pull request by zsarnett**
tap_action:
          action: call-service
          service: light.turn_off
          service_data:
            entity_id: light.office
        hold_action:
          action: call-service
          service: light.toggle
          service_data:
            entity_id: light.office

### CODE REVIEW

File: src/panels/lovelace/cards/hui-entity-button-card.ts [hunk 1]
```diff
@@ -23,15 +23,19 @@ import { LovelaceCardConfig } from "../../../data/lovelace";
 import { longPress } from "../common/directives/long-press-directive";
 import { fireEvent } from "../../../common/dom/fire_event";
 
+interface ActionConfig {
+  action?: "toggle" | "call-service" | "more-info";
```
Comment (iantrich) [inline src/panels/lovelace/cards/hui-entity-button-card.ts]: should we include `navigate` to better align our available actions?

**Comment by zsarnett** [comment]
UI Editor will need updated after this is merged

File: src/data/lovelace.ts [hunk 2]
```diff
@@ -21,6 +21,12 @@ export interface LovelaceCardConfig {
   [key: string]: any;
 }
 
+export interface ActionConfig {
+  action?: "toggle" | "call-service" | "more-info" | "none";
```
Comment (iantrich) [inline src/data/lovelace.ts]:
```suggestion
  action?: "toggle" | "call-service" | "more-info" | "navigate" | "none";
```

File: src/data/lovelace.ts [hunk 3]
```diff
@@ -21,6 +21,12 @@ export interface LovelaceCardConfig {
...
+  service?: string;
+  service_data?: object;
```
Comment (iantrich) [inline src/data/lovelace.ts]:
```suggestion
  service_data?: object;
  path?: string;
```

File: src/panels/lovelace/cards/hui-glance-card.ts [hunk 4]
```diff
@@ -243,24 +240,14 @@ export class HuiGlanceCard extends hassLocalizeLitMixin(LitElement)
     `;
   }
 
-  private handleClick(ev: MouseEvent, hold: boolean): void {
+  private _handleTap(ev) {
```
Comment (iantrich) [inline src/panels/lovelace/cards/hui-glance-card.ts]:
```suggestion
  private _handleTap(ev: MouseEvent) {
```

File: src/panels/lovelace/cards/hui-glance-card.ts [hunk 5]
```diff
@@ -243,24 +240,14 @@ export class HuiGlanceCard extends hassLocalizeLitMixin(LitElement)
...
-        break;
-      case "more-info":
-        fireEvent(this, "hass-more-info", { entityId });
-        break;
-      default:
-    }
+    handleClick(this, this.hass!, config, false);
+  }
+
+  private _handleHold(ev) {
```
Comment (iantrich) [inline src/panels/lovelace/cards/hui-glance-card.ts]:
```suggestion
  private _handleHold(ev: MouseEvent) {
```

File: src/panels/lovelace/common/handle-click.ts [hunk 6]
```diff
@@ -34,14 +45,15 @@ export const handleClick = (
...
-          entity_id: config.entity,
-          ...config.service_data,
-        };
-        hass.callService(domain, service, serviceData);
+      if (!actionConfig.service) {
+        return;
       }
+      const [domain, service] = actionConfig.service.split(".", 2);
+      const serviceData = {
+        entity_id: config.entity,
```
Comment (iantrich) [inline src/panels/lovelace/common/handle-click.ts]: This has caused issues for some services that don't want the `entity_id`. I can't recall them off the top of my head, but should require that the `service_data` be explicitly stated instead?

Comment (zsarnett) [reply src/panels/lovelace/common/handle-click.ts]: If it doesnt want it, the service should ignore it... I think that should be a service fix

**Comment by iantrich** [comment]
Travis threw an error here:
https://github.com/home-assistant/home-assistant-polymer/pull/2182/files#diff-2e86734b42da11910dbfae0763c5d602R77
Need to look inside the ActionConfigs now

File: src/panels/lovelace/common/handle-click.ts [hunk 7]
```diff
@@ -3,20 +3,30 @@ import { LovelaceElementConfig } from "../elements/types";
...
+  let actionConfig;
   if (hold && config.hold_action) {
-    action = config.hold_action;
+    actionConfig = config.hold_action;
+  } else if (!hold && config.tap_action) {
+    actionConfig = config.tap_action;
   }
 
+  const action =
+    actionConfig && actionConfig.action ? actionConfig.action : "more-info";
```
Comment (iantrich) [inline src/panels/lovelace/common/handle-click.ts]: Not all the types handled by `config` have an `action` field. I don't think any elements are likely working right now with this

Comment (zsarnett) [reply src/panels/lovelace/common/handle-click.ts]: Thats why it has the check for action field

Comment (iantrich) [reply src/panels/lovelace/common/handle-click.ts]: But you can still define an action type for the elements beyond "more-info". That is lost here.

Comment (iantrich) [reply src/panels/lovelace/common/handle-click.ts]: Perhaps we should just align the elements now as well? `icon`, `image`, `state-icon` and `state-label`

Comment (zsarnett) [reply src/panels/lovelace/common/handle-click.ts]: Oh you are saying the elements of picture elements
Yea I havent gotten there yet. Next commit. Looking at that now

File: src/panels/lovelace/common/handle-click.ts [hunk 8]
```diff
@@ -3,20 +3,34 @@ import { LovelaceElementConfig } from "../elements/types";
...
+  tap_action?: ActionConfig;
+  hold_action?: ActionConfig;
+}
 
 export const handleClick = (
   node: HTMLElement,
   hass: HomeAssistant,
-  config: LovelaceElementConfig | LovelaceCardConfig,
+  config:
+    | LovelaceElementConfig
```
Comment (iantrich) [inline src/panels/lovelace/common/handle-click.ts]: duplicate

File: src/panels/lovelace/elements/types.ts [hunk 9]
```diff
@@ -1,14 +1,14 @@
...
   type: string;
   style: object;
   entity?: string;
-  hold_action?: string;
-  navigation_path?: string;
+  hold_action?: ActionConfig;
   service?: string;
   service_data?: object;
-  tap_action?: string;
+  tap_action?: ActionConfig;
```
Comment (iantrich) [inline src/panels/lovelace/elements/types.ts]: Much like the `picture` card, in a future PR we should align `service-button` to use `tap_action` and `hold_action` so we can remove `service` and `service_data` from `LovelaceElementConfig`

**Review by iantrich — APPROVED** [review]

**Comment by zsarnett** [comment]
Correct. I though we decided against that. So that it was better configurable via the editor

**Comment by thomasloven** [comment]
Rereading the discussion it seems that way... but it was not what I thought I agreed to.

It looks much nicer in a text config, and is similar to the entity list, but as you say it'd have to be converted by the editor on first edit...

File: src/panels/lovelace/common/handle-click.ts (hunk 6 above)
Comment (balloob) [reply src/panels/lovelace/common/handle-click.ts]: Yeah we should stop trying to be smart and adding the entity automatically.

Comment (balloob) [reply src/panels/lovelace/common/handle-click.ts]: We can still automatically populate service data for `toggle`, but not for normal services.

**Comment by balloob** [comment]
I did think we would support the short-form too, when possible. However, given that editors are going to rewrite it and that by not defaulting service-data, `action: service` becomes semi-useless (but at least not unexpected behavior), we should probably just drop it.

I expect that within reasonable time, 95% of our users won't need to write YAML, especially since custom cards can also add config UI.

File: src/data/lovelace.ts [hunk 10]
```diff
@@ -21,6 +21,13 @@ export interface LovelaceCardConfig {
   [key: string]: any;
 }
 
+export interface ActionConfig {
```
Comment (balloob) [inline src/data/lovelace.ts]: This is the wrong approach. We should have a separate interface for each action and then have `ActionConfig` be a union interface: `ActionConfig = CallServiceActionConfig & MoreInfoActionConfig`

**Review by balloob — APPROVED** [review]

**Comment by iantrich** [comment]
fixes https://github.com/home-assistant/home-assistant-polymer/issues/2180

**Pull request merged**